lib/internal/inspector.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+
+const {
+  JSONParse,
+  JSONStringify,
+  Symbol,
+} = primordials;
+const EventEmitter = require('events');
+const assert = require('internal/assert');
+const {
+  ERR_INSPECTOR_ALREADY_CONNECTED,
+  ERR_INSPECTOR_COMMAND,
+  ERR_INSPECTOR_NOT_CONNECTED,
+  ERR_INVALID_ARG_TYPE,
+} = require('internal/errors').codes;
+const debug = require('internal/util/debuglog').debuglog('inspector');
+const { validateString } = require('internal/validators');
+
+const {
+  Connection,
+} = internalBinding('inspector');
+
+const connectionSymbol = Symbol('connectionProperty');
+const messageCallbacksSymbol = Symbol('messageCallbacks');
+const nextIdSymbol = Symbol('nextId');
+const onMessageSymbol = Symbol('onMessage');
+const resultSymbol = Symbol('result');
+
+// The synchronicity of this class depends on:
+// - ChannelImpl::sendProtocolResponse is synchronous
+// - InspectorSessionDelegate::sendMessageToFrontend is synchronous
+// - JSBindingsConnection::OnMessage is synchronous
+// when the Connection is instantiated with is_sync = true.
+class SyncSession extends EventEmitter {
+  constructor() {
+    super();
+    this[connectionSymbol] = null;
+    this[nextIdSymbol] = 1;
+    this[resultSymbol] = undefined;
+  }
+
+  connect() {
+    if (this[connectionSymbol])
+      throw new ERR_INSPECTOR_ALREADY_CONNECTED('The inspector session');
+    const connection =
+      new Connection(this[onMessageSymbol].bind(this), true);
+    if (connection.sessionAttached) {
+      throw new ERR_INSPECTOR_ALREADY_CONNECTED('Another inspector session');
+    }
+    this[connectionSymbol] = connection;
+  }
+
+  [onMessageSymbol](message) {
+    const parsed = JSONParse(message);
+    if (parsed.error) {
+      throw (new ERR_INSPECTOR_COMMAND(parsed.error.code,
+                                       parsed.error.message));
+    }
+    debug(`received message #${parsed.id}:`, parsed);
+    if (parsed.id) {
+      this[resultSymbol] = parsed.result;
+    } else {
+      this.emit(parsed.method, parsed);
+      this.emit('inspectorNotification', parsed);
+    }
+  }
+
+  post(method, params) {
+    validateString(method, 'method');
+    if (params && typeof params !== 'object') {
+      throw new ERR_INVALID_ARG_TYPE('params', 'Object', params);
+    }
+    if (!this[connectionSymbol]) {
+      throw new ERR_INSPECTOR_NOT_CONNECTED();
+    }
+    const id = this[nextIdSymbol]++;
+    const message = { id, method };
+    if (params) {
+      message.params = params;
+    }
+    this[resultSymbol] = undefined;
+    // [onMessageSymbol] is supposed to be called synchronously here
+    // to store the result in this[resultSymbol].
+    debug(`dispatching message #${id}:`, message);
+    this[connectionSymbol].dispatch(JSONStringify(message));
+    const result = this[resultSymbol];
```

Comment:
Note that implicit synchronicity is still there - it's not guaranteed by V8 that the result would show up as `this[resultSymbol]` right after the call to `this[connectionSymbol].dispatch(JSONStringify(message))`, before `post()` returns. It only happens to work right now.